Add deprecation decorator.

HideakiImamura opened the pull request:
## Motivation
This PR aims to provide the deprecation decorator as mentioned in Issue #1129. This PR is an alternative implementation of PR #1146.

## Description of the changes
- Add the deprecation decorator.
- Add the test for the deprecation decorator.

hvy commented on optuna/_deprecation.py:
```diff
@@ -0,0 +1,92 @@
+"""
+
+
+def deprecation(version: str, name: str = None) -> Any:
```
How about `deprecated`? It also follow the same format as `experimental`.

hvy commented on optuna/_deprecation.py:
```diff
@@ -0,0 +1,92 @@
+    """Decorate class or function as deprecation.
+
+    Args:
+        version: The first version that supports the target feature.
```
Seems like this text is copied from the experimental decorator but should probably be updated.

hvy commented on optuna/_deprecation.py:
```diff
@@ -0,0 +1,92 @@
+            indent = _get_docstring_indent(func.__doc__)
+            func.__doc__ = func.__doc__.strip() + textwrap.indent(note, indent) + indent
+
+            # TODO(crcrpar): Annotate this correctly.
```
A nit but probably best not to copy TODOs with names with developers. crcrpar is not necessarily aware/responsible for `_deprecation.py`. How about writing your own name?

hvy commented on optuna/_deprecation.py:
```diff
@@ -0,0 +1,92 @@
+            @functools.wraps(func)
+            def new_func(*args: Any, **kwargs: Any) -> Any:
+                warnings.warn(
+                    "{} is deprecated in v{}. "
```
Not sure but I think past tense sounds more natural.
```suggestion
                    "{} has been deprecated in v{}. "
```

toshihikoyanase commented on optuna/_deprecation.py:
```diff
@@ -0,0 +1,103 @@
+        name: The name of the feature. Defaults to the function or class name. Optional.
+    """
+
+    _validate_version(deprecated_version)
+    _validate_version(removed_version)
```
Maybe this is obvious, but do we add a check to ensure `deprecated_version < removed_version`? It may be able to prevent careless mistakes.

hvy commented:
> Related to #1382 (comment), do we also rename the filename to optuna/_deprecated.py?

+1. That seems reasonable and consistent with the experimental decorator.

> How do we call this decorator? I think we have two choices: 'deprecation decorator' (like 'deprecation warning') 'deprecated decorator' (like 'experimental decorator')?

Hm, I understand your concern. "deprecation decorator" and "experimental decorator" seem fine to me. Depends on how you see it but it's consistent with [`DeprecationWarning`](https://docs.python.org/3/library/exceptions.html#DeprecationWarning) and our [`ExperimentalWarning`](https://github.com/optuna/optuna/blob/master/optuna/exceptions.py#L79).

HideakiImamura commented:
@hvy, @toshihikoyanase Thanks for the review! I addressed all of your reviews. Could you take a look?

> "deprecation decorator" and "experimental decorator" seem fine to me. Depends on how you see it but it's consistent with `DeprecationWarning` and our `ExperimentalWarning`.

So, should we rename the current `deprecated` function to `deprecation`? I think it is reasonable to match the `experimental` decorator.

hvy commented:
>> "deprecation decorator" and "experimental decorator" seem fine to me. Depends on how you see it but it's consistent with DeprecationWarning and our ExperimentalWarning.

> So, should we rename the current deprecated function to deprecation? I think it is reasonable to match the experimental decorator.

I did not mean to rename the decorator to `deprecation`. I was talking about how to refer to this decorator in plain speech/text.

hvy commented on tests/test_deprecation.py:
```diff
@@ -0,0 +1,114 @@
+from typing import Any
```
How about renaming this too to `test_deprecated`?

hvy commented:
Sorry for my late comment but to make this decorator useful, we should probably add another argument for appending arbitrary text to the notes. If you look at existing deprecation warnings in Optuna, e.g. CLI, `best_booster` and `structs`, they all forward the user to some alternative API to use instead of the deprecated one. It's not obvious how we should include these messages with the current approach.

toshihikoyanase commented:
> I did not mean to rename the decorator to deprecation. I was talking about how to refer to this decorator in plain speech/text.

Thank you for your elaboration. I didn't want to rename the decorator too. I found that the some test cases use 'deprecation decorator' in the function names, so I just wondered how do we call it in conversation.

> we should probably add another argument for appending arbitrary text to the notes.

That makes sense. It definitely helps users' migration.

hvy commented on optuna/_deprecated.py:
```diff
@@ -0,0 +1,138 @@
+            note = (
+                _DEPRECATION_NOTE_TEMPLATE.format(d_ver=deprecated_version, r_ver=removed_version)
+                if deprecation_note is None
+                else _formatting_note(deprecation_note)
```
Is it difficult to append the `deprecation_note` to the `_DEPRECATION_NOTE_TEMPLATE`, or do you prefer this way using either? I think it's simpler always keeping the template, thinking of this note as an optional body after it.

hvy commented on optuna/_deprecated.py:
```diff
@@ -0,0 +1,138 @@
+
+            .. note::
+                The specified text must be separated by the newline characters, and each line
+                should not contain more than 92 characters without the added newline character.
```
Mostly curious but where does this 92 come from?

HideakiImamura replied:
I initially thought that we need to bound the length of each line <100, but isn't that necessary?

crcrpar commented on optuna/_deprecated.py:
```diff
@@ -0,0 +1,141 @@
+
+
+def deprecated(
+    deprecated_version: str, removed_version: str, name: str = None, additional_text: str = ""
```
[nit-picking]
I guess most of type hintings come from my terrible `experimental` implementation, though, if you allow `None` explicitly for `name` argument, then it should be `Optional[str]`. Personally empty string is enough as the default for `name`.
```
    deprecated_version: str, removed_version: str, name: Optional[str] = None, additional_text: str = ""
```

or 

```
    deprecated_version: str, removed_version: str, name: str = "", additional_text: str = ""
```

HideakiImamura replied:
Thanks a lot! Fixed.

harupy commented:
@hvy @toshihikoyanase @HideakiImamura 

How about using the `.. warning::` directive instead of `.. note::` to differentiate deprecated warning blocks? There is `.. deprecated::` directive, but `sphinx-rtd-theme` doesn't seem to highlight it.

https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#paragraph-level-markup

![Screen Shot 2020-06-22 at 20 34 38](https://user-images.githubusercontent.com/17039389/85283535-7d7ca100-b4c8-11ea-9a83-c3843fa353d7.png)

Source:

```python
_DEPRECATION_NOTE_TEMPLATE = """

.. warning::
    Deprecated in v{d_ver}. This feature will be removed in the future. The removal of this
    feature is currently scheduled for v{r_ver}, but this schedule is subject to change.
    See https://github.com/optuna/optuna/releases/tag/v{d_ver}.

.. note::
    Deprecated in v{d_ver}. This feature will be removed in the future. The removal of this
    feature is currently scheduled for v{r_ver}, but this schedule is subject to change.
    See https://github.com/optuna/optuna/releases/tag/v{d_ver}.

.. deprecated::
    Deprecated in v{d_ver}. This feature will be removed in the future. The removal of this
    feature is currently scheduled for v{r_ver}, but this schedule is subject to change.
    See https://github.com/optuna/optuna/releases/tag/v{d_ver}.
"""
```

hvy commented:
That sounds like a good idea. A deprecation (`DeprecationWarning`) might be more of a warning than a "note".

hvy commented:
Fixed by #1407.
